Add output buffer ad5686 master

mirceacaprioru opened the pull request:
This set of patches adds output buffer support for the AD5686 family of nano DACs. It uses a PWM source to trigger the interrupt to read samples from the buffer and output on DAC channels.

dbogdan commented on drivers/iio/dac/ad5686.h:
```diff
@@ -88,6 +89,12 @@ enum ad5686_regmap_type {
+enum ad5686_iio_dev_attr {
+	DUTY,
+	PERIOD,
+	ENABLE
```
Shouldn't the driver only expose the sampling rate information and automatically calculate the duty and the period?

mirceacaprioru replied:
The duty cycle doesn't make any difference here since we use the rising edge anyway, i could convert to sampling frequency and have the duty set to a default 50%.

commodo commented:
checkpatch has one complaint:
https://travis-ci.org/github/analogdevicesinc/linux/jobs/713219795#L373

looks valid

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -25,6 +25,111 @@ static const char * const ad5686_powerdown_modes[] = {
+	}
+
+	pwm_set_relative_duty_cycle(&state, st->pwm_duty, 100);
+	ret |= pwm_apply_state(st->pwm, &state);
```
why `ret |=` ?  do you still need to continue if `ret` has an error?

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -25,6 +25,111 @@ static const char * const ad5686_powerdown_modes[] = {
+		ret = -EINVAL;
+	}
+
+	return ret;
```
it seems to make sense to make all these return statement in the switch() statement;

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -25,6 +25,111 @@ static const char * const ad5686_powerdown_modes[] = {
+			ad5686_store,
+			ENABLE);
+
+static struct attribute *ad5686_attribute[] = {
```
does the PWM framework has it's own sysfs?
to avoid doing one for the driver?

mirceacaprioru replied:
it has a minimal sysfs but you have the export the device manually: https://www.kernel.org/doc/html/latest/driver-api/pwm.html#using-pwms-with-the-sysfs-interface
I would like to have a direct pwm control through the driver to increase/decrease sampling frequency.

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -533,11 +539,47 @@ static const struct ad5686_chip_info ad5686_chip_info_tbl[] = {
+		goto out;
+
+	for_each_set_bit(i, indio_dev->active_scan_mask, indio_dev->masklength) {
+
```
nitpick: extra line;

commodo replied:
using `active_scan_mask` & `masklength` are not recommended; they are internal fields of the IIO framework;
is there some other way to do this without using these?

mirceacaprioru replied:
this is something i picked up from ad7303: https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/dac/ad7303.c#L78
i could try to use something else will look into more details.

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -533,11 +539,47 @@ static const struct ad5686_chip_info ad5686_chip_info_tbl[] = {
+		chan = iio_find_channel_from_si(indio_dev, i);
+		ret = st->write(st, AD5686_CMD_WRITE_INPUT_N_UPDATE_N,
+				chan->address, val << chan->scan_type.shift);
+		mutex_unlock(&indio_dev->mlock);
```
mlock is also framework-only

mirceacaprioru replied:
the driver should be shifted to it's own lock so i will also do that modification.

mirceacaprioru commented:
Changelog v2
- moved driver from using indio_dev->mlock to its own lock
- translated pwm_period to sampling frequency
- added trgger enable function to use it automatically when the buffer is enabled this will also start the PWM generation

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -127,9 +127,9 @@ static int ad5686_read_raw(struct iio_dev *indio_dev,
 	switch (m) {
 	case IIO_CHAN_INFO_RAW:
-		mutex_lock(&indio_dev->mlock);
+		mutex_lock(&st->lock);
```
you can also backport this: https://github.com/torvalds/linux/commit/0b2884ef7fcad881556b79828a96f359c7f1fab6
whatever works;

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -472,6 +488,20 @@ int ad5686_probe(struct device *dev,
 	}
 
+	/* PWM configuration */
+	st->pwm = devm_pwm_get(dev, NULL);
```
just curios: is there a chance we would need another PWM? to name this now;

mirceacaprioru replied:
only if we go with the multi buffer approach that you are developing for input buffers, but it makes sense to be more precise in aquiring components. will add

commodo commented on drivers/iio/dac/ad5686.c:
```diff
@@ -442,11 +478,46 @@ static const struct ad5686_chip_info ad5686_chip_info_tbl[] = {
+	for_each_set_bit(i, indio_dev->active_scan_mask, indio_dev->masklength) {
+		val = (sample[1] << 8) + sample[0];
+
+		mutex_lock(&st->lock);
```
would it make sense to keep the lock held for the entire loop?

mirceacaprioru replied:
good point will move the mutex out

commodo commented:
rest looks good

mirceacaprioru commented:
Changelog V3
- add pwm name
- move mutex out of for loop
- remove lock commit since the upstream version was backported
